### PR TITLE
bpo-34160: Preserves order of minidom of Element attributes

### PR DESCRIPTION
**Pull request by dfrojas**
This PR is a continuation of the work in https://github.com/python/cpython/pull/10163

https://bugs.python.org/issue34160

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Lib/xml/etree/ElementTree.py [hunk 1]
```diff
@@ -979,7 +979,7 @@ def _serialize_html(write, elem, qnames, namespaces, **kwargs):
                             k,
                             _escape_attrib(v)
                             ))
-                for k, v in sorted(items):  # lexical order
+                for k, v in items:
```
Comment (vstinner) [inline Lib/xml/etree/ElementTree.py]: This change has already been committed: commit 3b05ad7be09af1d4510eb698b0a70d36387f296e, PR #10190.

File: Lib/xml/dom/minidom.py [hunk 2]
```diff
@@ -854,7 +854,7 @@ def writexml(self, writer, indent="", addindent="", newl=""):
         writer.write(indent+"<" + self.tagName)
 
         attrs = self._get_attributes()
-        a_names = sorted(attrs.keys())
+        a_names = attrs.keys()
 
         for a_name in a_names:
```
Comment (vstinner) [inline Lib/xml/dom/minidom.py]:
```suggestion
        for a_name in attrs.keys():
```

File: Doc/library/xml.etree.elementtree.rst [hunk 3]
```diff
@@ -642,6 +642,13 @@ Functions
    :class:`Element` instance and a dictionary.
 
 
+.. function:: _serialize_html(self, writer, indent="", addindent="", newl="")
```
Comment (vstinner) [inline Doc/library/xml.etree.elementtree.rst]: No, please don't document private functions.

File: Doc/library/xml.dom.minidom.rst [hunk 4]
```diff
@@ -143,6 +143,9 @@ module documentation.  This section lists the differences between the API and
    For the :class:`Document` node, an additional keyword argument *encoding* can
    be used to specify the encoding field of the XML header.
 
+   .. versionchanged:: 3.8
+        The :meth:`writexml` method now preserves the attribute order specified
```
Comment (serhiy-storchaka) [inline Doc/library/xml.dom.minidom.rst]: Wrong indentation.

File: Lib/test/test_minidom.py [hunk 5]
```diff
@@ -1560,5 +1561,21 @@ def testProcessingInstructionNameError(self):
...
 
+    def test_minidom_attribute_order(self):
+        doc = Document()
+        root = doc.createElement("curriculum")
+        root.setAttribute( "status", 'public' )
+        root.setAttribute( "company", 'example' )
+        doc.appendChild(root)
+
+        with open(support.TESTFN, "w") as xml_file:
+            doc.writexml(xml_file)
```
Comment (serhiy-storchaka) [inline Lib/test/test_minidom.py]: I am surprised, but `writexml()` was not tested before. `toxml()` and `toprettyxml()` are tested. Please add tests also for these methods.

File: Lib/test/test_minidom.py [hunk 6]
```diff
@@ -1560,5 +1561,21 @@ def testProcessingInstructionNameError(self):
...
         pi.nodeValue = "crash"
 
+    def test_minidom_attribute_order(self):
+        doc = Document()
+        root = doc.createElement("curriculum")
+        root.setAttribute( "status", 'public' )
+        root.setAttribute( "company", 'example' )
+        doc.appendChild(root)
+
+        with open(support.TESTFN, "w") as xml_file:
```
Comment (serhiy-storchaka) [inline Lib/test/test_minidom.py]: Don't use `open()` in text mode without explicit encoding in tests (unless you test `open()` itself). The result is environment depending. It is better to use `io.StringIO` in this test.

File: Lib/test/test_minidom.py [hunk 7]
```diff
@@ -1560,5 +1561,21 @@ def testProcessingInstructionNameError(self):
         pi = doc.createProcessingInstruction("y", "z")
         pi.nodeValue = "crash"
 
+    def test_minidom_attribute_order(self):
+        doc = Document()
```
Comment (serhiy-storchaka) [inline Lib/test/test_minidom.py]: It may be easier to use `parseString()`.

File: Lib/test/test_xml_etree.py [hunk 8]
```diff
@@ -1064,6 +1064,16 @@ def test_tree_write_attribute_order(self):
                          b"<?xml version='1.0' encoding='utf-8'?>\n"
                          b'<cirriculum status="public" company="example" />')
 
+    def test_html_serialize_attribute_order(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py]: This is already tested.

File: Doc/library/xml.dom.minidom.rst (hunk 4 above)
Comment (serhiy-storchaka) [reply Doc/library/xml.dom.minidom.rst]: Document also this for `toxml()` and `toprettyxml()`.

File: Lib/xml/etree/ElementTree.py (hunk 1 above)
Comment (dfrojas) [reply Lib/xml/etree/ElementTree.py]: You are right, I did not realize that @serhiy-storchaka have made that changes some minutes before this PR.

File: Lib/test/test_minidom.py (hunk 5 above)
Comment (dfrojas) [reply Lib/test/test_minidom.py]: Done

File: Lib/test/test_minidom.py (hunk 6 above)
Comment (dfrojas) [reply Lib/test/test_minidom.py]: You are right, done. 

I just saw your comment in the @rhettinger PR saying the same

**Comment by dfrojas** [comment]
@vstinner @serhiy-storchaka I had conflicts doing the changes, the PR has commits from other developers. It's better if I close this PR and open a new one?

**Comment by serhiy-storchaka** [comment]
Something wrong happen when you synchronized your branch with master. Did you run `git merge master`? If there are conflicts, it is better to resolve them and continue in the same PR.

**Comment by taleinat** [comment]
@dfrojas, I recommend:

1. Create a new branch from the current master branch.
2. Cherry-pick your relevant commits one by one in chronological order.
3. Review the changes, run the tests etc.
4. If everything looks okay, force-push the new branch into the branch for this PR, e.g. on the new branch: `git push -f origin bpo-34160-10163PR-continuation`

File: Lib/test/test_minidom.py [hunk 9]
```diff
@@ -1560,5 +1562,28 @@ def testProcessingInstructionNameError(self):
         pi = doc.createProcessingInstruction("y", "z")
         pi.nodeValue = "crash"
 
+    def test_minidom_attribute_order(self):
+        xml_str = '<?xml version="1.0" ?><curriculum status="public" company="example"/>'
+        doc = parseString(xml_str)
+        output = io.StringIO()
+
+        with contextlib.redirect_stdout(output) as xml_file:
```
Comment (serhiy-storchaka) [inline Lib/test/test_minidom.py]: `output` and `xml_file` are the same. You can either remove `as xml_file` or replace `output` with `io.StringIO()`.

File: Lib/test/test_minidom.py [hunk 10]
```diff
@@ -1560,5 +1562,28 @@ def testProcessingInstructionNameError(self):
...
+            doc.writexml(xml_file)
+        self.assertEqual(output.getvalue(),
+                         '<?xml version="1.0" ?><curriculum status="public" company="example"/>')
```
Comment (serhiy-storchaka) [inline Lib/test/test_minidom.py]: Is it `xml_str`? Maybe use this variable?

Comment (dfrojas) [reply Lib/test/test_minidom.py]: You are right, is redundant. I thought to do it with a different object to be sure of the output.

File: Lib/test/test_minidom.py (hunk 9 above)
Comment (dfrojas) [reply Lib/test/test_minidom.py]: Done.

**Comment by dfrojas** [comment]
@serhiy-storchaka  Do I have to write a NEWS specifically for minidom? Or with the NEWS in [this PR ](https://github.com/python/cpython/pull/10163/files) is enough?

**Comment by taleinat** [comment]
> 
> 
> @serhiy-storchaka Do I have to write a NEWS specifically for minidom? Or with the NEWS in [this PR ](https://github.com/python/cpython/pull/10163/files) is enough?

Perhaps just update the NEWS entry you referenced and add "minidom" after "ElementTree".

**Comment by serhiy-storchaka** [comment]
Either add a new news entry, or update an existing entry added for #10219. The rest LGTM.

**Review by serhiy-storchaka — APPROVED** [review]

File: Misc/NEWS.d/next/Library/2018-10-27-21-11-42.bpo-34160.UzyPZf.rst [hunk 11]
```diff
@@ -1 +1 @@
-ElementTree now preserves the attribute order specified by the user.
+ElementTree and minidom now preserves the attribute order specified by the user.
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-10-27-21-11-42.bpo-34160.UzyPZf.rst]:
```suggestion
ElementTree and minidom now preserve the attribute order specified by the user.
```

Comment (dfrojas) [reply Misc/NEWS.d/next/Library/2018-10-27-21-11-42.bpo-34160.UzyPZf.rst]: Done.

**Pull request merged**